src/main.rs

```diff
@@ -259,25 +329,44 @@ impl App {
             None
         }
     }
+    fn handle_hover(&self, params: TextDocumentPositionParams) -> Option<(Option<Range>, String)> {
+        let (_, content, tree) = self.files.get(&params.text_document.uri)?;
+        let offset = utils::lookup_pos(content, params.position)?;
+        let tree = match tree {
+            Ok(x) => {
+                let tmp = Gc::new(x.clone());
+                climb_tree(&tmp, offset).clone()
+            }
+            Err(e) => return Some((None, format!("{:?}", e))),
+        };
+        let range = utils::range(content, tree.range?);
+        let val = match tree.eval() {
+            Ok(x) => x.format_markdown(),
+            Err(e) => format!("{}", e),
+        };
+        Some((Some(range), val))
+    }
```

Comment:
Now that we can load attr keys here as well: should we use this procedure for the completions as well?

Reply:
Pretty much. In `nix-eval-lsp`, the only major difference is some code at the end pushing completions to a list:

https://github.com/aaronjanse/nix-eval-lsp/blob/0c263b802d83b3988924d6db55a6a0ca60faf578/src/main.rs#L235

Note that we'll want to merge rnix-lsp logic with evaluator logic to allow graceful fallback to static analysis when the evaluator gets confused.